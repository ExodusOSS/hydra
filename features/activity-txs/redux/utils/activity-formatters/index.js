import { formatSwapActivity } from './format-swap-activity'
import { formatTxActivity } from './format-tx-activity'

export const formattersByType = new Map([
  ['exchange', formatSwapActivity],
  ['tx', formatTxActivity],
])
