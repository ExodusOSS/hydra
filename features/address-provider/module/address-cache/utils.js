import lodash from 'lodash'
import assert from 'minimalistic-assert'
import { set } from '@exodus/basic-utils'

const { get } = lodash

// This allows new assets to use a full key, without regenerating existing data in fusion for legacy ones. Please do NOT add new names here!
const legacyKeyAssetNames = new Set([
  'algorand',
  'ark',
  'avalanchec',
  'bcash',
  'bgold',
  'bitcoin',
  'bitcoinsv',
  'bnbmainnet',
  'bsc',
  'cardano',
  'cosmos',
  'dash',
  'decred',
  'digibyte',
  'dogecoin',
  'elrond',
  'eosio',
  'ethereum',
  'ethereumarbnova',
  'ethereumarbone',
  'ethereumclassic',
  'ethereumgoerli',
  'fantommainnet',
  'filecoin',
  'flare',
  'harmonymainnet',
  'hedera',
  'iconmainnet',
  'lightningnetwork',
  'lisk',
  'litecoin',
  'matic',
  'neo3',
  'monero',
  'nano',
  'nem',
  'neo',
  'ontology',
  'optimism',
  'polkadot',
  'qtumignition',
  'ravencoin',
  'ripple',
  'solana',
  'stellar',
  'terra',
  'tezos',
  'theta',
  'tronmainnet',
  'vechainthor',
  'vertcoin',
  'waves',
  'zcash',
  'zilliqa',
])

export const getCachePath = ({ walletAccountName, baseAssetName, derivationPath }) => {
  assert(typeof walletAccountName === 'string', 'expected string "walletAccountName"')
  assert(typeof baseAssetName === 'string', 'expected string "baseAssetName"')
  assert(typeof derivationPath === 'string', 'expected string "derivationPath"')

  // todo add assetName for all assets when we migrate to a new fusion channel
  let path = derivationPath
  if (!legacyKeyAssetNames.has(baseAssetName)) {
    path = `${baseAssetName}/${path}` // combine these two to match fusion shape
  }

  return [
    // walletAccount is indirectly captured in the BIP path but it is still required because the BIP path is not enough, Trezor's accountIndex always use 1 regardless of the index of a walletAccount.
    walletAccountName,
    // assetName is required because some assets encode the network identifier into the address e.g splitting mainnet vs testnet results in different addresses for the same public key.
    // bipPath is required for assets that may have multiple addresses, or stake addresses etc
    path,
  ]
}

export const diffCaches = (cache1, cache2, fromSync) => {
  let isDifferent = false
  let needsSync = false

  const result = Object.create(null)
  for (const walletAccount in cache2) {
    const subCache = cache2[walletAccount]
    for (const path in subCache) {
      const existing = get(cache1, [walletAccount, path])
      const incoming = subCache[path]
      const newAddress = !existing
      const addressMismatch = !newAddress && existing.address !== incoming.address
      needsSync = needsSync || existing?.synced === false
      const justSynced = !existing?.synced && incoming.synced
      if (fromSync && addressMismatch && !existing?.synced) {
        // We have an address scheduled to sync up, that is mismatching with a
        // sync down, in this case we actually prefer the one scheduled to sync up.
        set(result, [walletAccount, path], existing)
      } else if (newAddress || addressMismatch || justSynced) {
        isDifferent = true
        set(result, [walletAccount, path], incoming)
      }
    }
  }

  return {
    diff: result,
    isDifferent,
    needsSync: isDifferent || needsSync,
  }
}

export const getUnsyncedAddressesForPush = (cache, currentCache) => {
  const out = Object.create(null)
  for (const walletAccount in cache) {
    const subCache = cache[walletAccount]
    for (const path in subCache) {
      const item = subCache[path]
      const currentCacheItem = get(currentCache, [walletAccount, path])
      if (
        currentCacheItem &&
        currentCacheItem.synced &&
        currentCacheItem.address.toString() === item.address.toString()
      )
        continue
      if (!item.synced) set(out, [walletAccount, path], item.address)
    }
  }

  return out
}
