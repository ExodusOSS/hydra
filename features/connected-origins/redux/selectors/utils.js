export const parseOrigin = (origin) =>
  origin
    .replace(/((^\w+:|^)\/\/)?(www\.)?/u, '')
    .replace(/\/$/u, '')
    .split('.')
    .slice(-2)
    .join('.')
