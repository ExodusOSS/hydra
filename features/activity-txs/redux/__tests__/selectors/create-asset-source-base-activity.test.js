import { omit } from '@exodus/basic-utils'
import { OrderSet, Tx } from '@exodus/models'

import { setup } from '../utils'

describe('createAssetSourceBaseActivity', () => {
  test('should return array of activities in wallet', () => {
    const { store, selectors } = setup()
    const empty = selectors.activityTxs.createAssetSourceBaseActivity({
      assetName: 'ethereum',
      walletAccount: 'exodus_0',
    })

    expect(empty(store.getState())).toEqual([])
  })

  test('should group txs with same order', () => {
    const { store, selectors, handleEvent, assets } = setup()

    const BTC_ETH_ORDER = {
      orderId: 'btc-eth-order',
      fromAsset: 'bitcoin',
      toAsset: 'ethereum',
      txIds: [{ txId: 'some-receive-tx' }, { txId: 'some-receive-tx-2' }],
      date: '2019-07-22T13:54:28.000Z',
      fromWalletAccount: 'exodus_0',
      toWalletAccount: 'exodus_0',
      toAmount: assets.ethereum.currency.defaultUnit(3),
    }

    const txLogFixtures = {
      ethereum: [
        {
          txId: 'some-receive-tx',
          error: null,
          date: '2019-07-22T13:54:28.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '1 ETH',
          coinName: 'ethereum',
          feeCoinName: 'ethereum',
          feeAmount: '0.000189 ETH',
          currencies: { ethereum: assets.ethereum.currency },
        },
        {
          txId: 'some-receive-tx-2',
          error: null,
          date: '2019-07-22T13:54:28.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '2 ETH',
          coinName: 'ethereum',
          feeCoinName: 'ethereum',
          feeAmount: '0.000189 ETH',
          currencies: { ethereum: assets.ethereum.currency },
        },
      ],
    }

    handleEvent('activityTxs', {
      exodus_0: {
        ethereum: txLogFixtures.ethereum.map((tx) => Tx.fromJSON(tx)),
      },
    })
    handleEvent('orders', OrderSet.fromArray([BTC_ETH_ORDER]))

    const selector = selectors.activityTxs.createAssetSourceBaseActivity({
      assetName: 'ethereum',
      walletAccount: 'exodus_0',
    })

    const result = selector(store.getState())

    expect(result.length).toEqual(1)
    expect(result[0].id).toEqual('swap.ethereum.btc-eth-order')
    expect(result[0].type).toEqual('exchange')
    expect(result).toMatchSnapshot()
  })

  test('should return activity for asset', () => {
    const { store, selectors, handleEvent, assets } = setup()

    const txLogFixtures = {
      ethereum: [
        {
          txId: 'some-receive-tx',
          error: null,
          date: '2020-07-22T13:54:28.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '1 ETH',
          coinName: 'ethereum',
          feeCoinName: 'ethereum',
          feeAmount: '0.000189 ETH',
          currencies: { ethereum: assets.ethereum.currency },
        },
        {
          txId: 'some-receive-tx-2',
          error: null,
          date: '2019-07-22T13:54:28.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '2 ETH',
          coinName: 'ethereum',
          feeCoinName: 'ethereum',
          feeAmount: '0.000189 ETH',
          currencies: { ethereum: assets.ethereum.currency },
        },
      ],
    }

    handleEvent('activityTxs', {
      exodus_0: {
        ethereum: txLogFixtures.ethereum.map((tx) => Tx.fromJSON(tx)),
      },
    })

    const selector = selectors.activityTxs.createAssetSourceBaseActivity({
      assetName: 'ethereum',
      walletAccount: 'exodus_0',
    })

    const result = selector(store.getState())

    expect(result.length).toEqual(2)
    expect(result[0].id).toEqual('ethereum.some-receive-tx')
    expect(result[1].id).toEqual('ethereum.some-receive-tx-2')
    expect(result).toMatchSnapshot()
  })

  test('should use activityIndex for id when available', () => {
    const { store, selectors, handleEvent, assets } = setup()

    const txLogFixtures = {
      bitcoin: [
        {
          currencies: { bitcoin: assets.bitcoin.currency },
          txId: 'bitcoin-tx',
          error: null,
          date: '2019-07-18T12:48:15.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '-0.1 BTC',
          coinName: 'bitcoin',
          feeCoinName: 'bitcoin',
          feeAmount: '0.00002 BTC',
          data: {
            activityIndex: 0,
            sent: [
              { amount: '0.1 BTC', address: 'btc-address-1' },
              { amount: '0.1 BTC', address: 'btc-address-2' },
            ],
          },
        },
        {
          currencies: { bitcoin: assets.bitcoin.currency },
          txId: 'bitcoin-tx',
          error: null,
          date: '2019-07-18T12:48:15.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '-0.1 BTC',
          coinName: 'bitcoin',
          feeCoinName: 'bitcoin',
          feeAmount: '0.00002 BTC',
          data: {
            activityIndex: 1,
            sent: [
              { amount: '0.1 BTC', address: 'btc-address-1' },
              { amount: '0.1 BTC', address: 'btc-address-2' },
            ],
          },
        },
      ],
    }

    handleEvent('activityTxs', {
      exodus_0: {
        bitcoin: txLogFixtures.bitcoin.map((tx) => Tx.fromJSON(tx)),
      },
    })

    const selector = selectors.activityTxs.createAssetSourceBaseActivity({
      assetName: 'bitcoin',
      walletAccount: 'exodus_0',
    })

    const result = selector(store.getState())

    expect(result.length).toEqual(2)
    expect(result[0].id).toEqual('bitcoin.bitcoin-tx')
    expect(result[1].id).toEqual('bitcoin.bitcoin-tx_1')
    expect(result).toMatchSnapshot()
  })

  test('should add indexless orders to activity', () => {
    const { store, selectors, handleEvent, assets } = setup()
    handleEvent('assets', {
      assets: {
        algorand: assets.algorand,
        flare: {
          ...assets.flare,
          baseAsset: {
            ...assets.flare.baseAsset,
            api: {
              features: {
                noHistory: true, // simulate flare have no history
              },
            },
          },
        },
      },
    })
    handleEvent('availableAssetNames', ['algorand', 'flare'])
    const orders = OrderSet.fromArray([
      {
        orderId: 'eac6bec7798be8',
        status: 'complete-verified',
        date: '2023-09-13T11:59:24.204Z',
        fromTxId: 'ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ',
        toTxId: '0x8bbea887472df32167597c36fffc70ac508a01088194e30b877d093fe25f9725',
        txIds: [
          {
            txType: 'depositTransaction',
            txId: 'ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ',
          },
          {
            txType: 'payoutTransaction',
            txId: '0x8bbea887472df32167597c36fffc70ac508a01088194e30b877d093fe25f9725',
          },
        ],
        fromAsset: 'algorand',
        fromAmount: {
          t: 'numberunit',
          v: {
            v: '1013.41878 ALGO',
            u: {
              microAlgo: 0,
              ALGO: 6,
            },
          },
        },
        toAsset: 'flare',
        toAmount: {
          t: 'numberunit',
          v: {
            v: '7598.14749525 FLR',
            u: {
              wei: 0,
              Kwei: 3,
              Mwei: 6,
              Gwei: 9,
              szabo: 12,
              finney: 15,
              FLR: 18,
            },
          },
        },
        fromWalletAccount: 'exodus_0',
        toWalletAccount: 'exodus_0',
        svc: 'cn',
        synced: true,
        region: 'INTL',
        displaySvc: 'cn',
        potentialToTxIds: [],
        _version: 1,
      },
    ])

    handleEvent('orders', orders)

    handleEvent('activityTxs', {
      exodus_0: {
        flare: [
          // Simulate having tx. it must be ignored
          {
            txId: '0x8bbea887472df32167597c36fffc70ac508a01088194e30b877d093fe25f9725',
            coinName: 'flare',
            coinAmount: '7598.14749525 FLR',
            confirmations: 1,
            date: '2023-09-13T11:59:24.204Z',
            currencies: { flare: assets.flare.currency },
          },
        ].map((tx) => Tx.fromJSON(tx)),
        algorand: [
          {
            txId: 'ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ',
            coinName: 'algorand',
            coinAmount: '1013.41878 ALGO',
            confirmations: 1,
            date: '2023-09-13T11:59:24.204Z',
            currencies: { algorand: assets.algorand.currency },
          },
        ].map((tx) => Tx.fromJSON(tx)),
      },
    })

    const selector = selectors.activityTxs.createAssetSourceBaseActivity({
      assetName: 'flare',
      walletAccount: 'exodus_0',
    })

    const result = selector(store.getState())
    expect(result.length).toEqual(1)
    expect(omit(result[0], ['toAsset', 'fromAsset'])).toMatchSnapshot()

    const selector2 = selectors.activityTxs.createAssetSourceBaseActivity({
      assetName: 'algorand',
      walletAccount: 'exodus_0',
    })
    const result2 = selector2(store.getState())
    expect(result2.length).toEqual(1)
    expect(omit(result2[0], ['toAsset', 'fromAsset'])).toMatchSnapshot()
  })

  test('should add orders without txs to activity', () => {
    const { store, selectors, handleEvent, assets } = setup()
    handleEvent('assets', {
      assets: {
        algorand: assets.algorand,
        flare: assets.flare,
      },
    })
    handleEvent('availableAssetNames', ['algorand', 'flare'])

    // simulate pending order without payout tx (received tx for flare)
    const orders = OrderSet.fromArray([
      // normal order with tx
      {
        orderId: 'order-with-tx',
        status: 'complete-verified',
        date: new Date('2022-09-13T11:59:24.204Z'),
        fromTxId: 'tx-with-order-algorand',
        toTxId: 'tx-with-order-flare',
        fromAsset: 'algorand',
        fromAmount: assets.algorand.currency.defaultUnit(1),
        toAsset: 'flare',
        toAmount: assets.flare.currency.defaultUnit(10),
        fromWalletAccount: 'exodus_0',
        toWalletAccount: 'exodus_0',
        svc: 'cn',
        synced: true,
        region: 'INTL',
        displaySvc: 'cn',
        potentialToTxIds: [],
        _version: 1,
      },
      // pending order without received tx
      {
        orderId: 'eac6bec7798be8',
        status: 'in-progress',
        date: new Date('2023-09-13T11:59:24.204Z'),
        fromTxId: 'algorand-send-tx',
        toTxId: null,
        fromAsset: 'algorand',
        fromAmount: assets.algorand.currency.defaultUnit(100),
        toAsset: 'flare',
        toAmount: assets.flare.currency.defaultUnit(10_000),
        fromWalletAccount: 'exodus_0',
        toWalletAccount: 'exodus_0',
        svc: 'cn',
        synced: true,
        region: 'INTL',
        displaySvc: 'cn',
        potentialToTxIds: [],
        _version: 1,
      },
    ])

    handleEvent('orders', orders)

    handleEvent('activityTxs', {
      exodus_0: {
        flare: [
          {
            txId: 'tx-with-order-flare',
            coinName: 'flare',
            coinAmount: '10 FLR',
            confirmations: 1,
            date: '2022-09-13T11:59:24.204Z',
            currencies: { flare: assets.flare.currency },
          },
        ].map((tx) => Tx.fromJSON(tx)),
        algorand: [
          {
            txId: 'tx-with-order-algorand',
            coinName: 'algorand',
            coinAmount: '1 ALGO',
            confirmations: 1,
            date: '2022-09-13T11:59:24.204Z',
            currencies: { algorand: assets.algorand.currency },
          },
          {
            txId: 'ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ',
            coinName: 'algorand',
            coinAmount: '1013.41878 ALGO',
            confirmations: 1,
            date: '2023-09-13T11:59:24.204Z',
            currencies: { algorand: assets.algorand.currency },
          },
        ].map((tx) => Tx.fromJSON(tx)),
      },
    })

    const selector = selectors.activityTxs.createAssetSourceBaseActivity({
      assetName: 'flare',
      walletAccount: 'exodus_0',
      displayOrderWithoutTx: true,
    })

    const result = selector(store.getState())
    expect(result.length).toEqual(2)
    expect(result[0].id).toEqual('swap.flare.eac6bec7798be8')
    expect(result[0].pending).toEqual(true)
    expect(result[1].id).toEqual('swap.flare.order-with-tx')
    expect(result[1].order.orderId).toEqual('order-with-tx')
    expect(result[1].pending).toEqual(false)
    expect(omit(result[0], ['toAsset', 'fromAsset'])).toMatchSnapshot()
    expect(omit(result[1], ['toAsset', 'fromAsset'])).toMatchSnapshot()
  })

  test('should return txs if orders are not supported', () => {
    const { store, selectors, handleEvent, assets } = setup({ withOrders: false })
    handleEvent('assets', {
      assets: {
        algorand: assets.algorand,
      },
    })
    handleEvent('availableAssetNames', ['algorand'])

    handleEvent('activityTxs', {
      exodus_0: {
        algorand: [
          {
            txId: 'ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ',
            coinName: 'algorand',
            coinAmount: '1013.41878 ALGO',
            confirmations: 1,
            date: '2023-09-13T11:59:24.204Z',
            currencies: { algorand: assets.algorand.currency },
          },
        ].map((tx) => Tx.fromJSON(tx)),
      },
    })

    const selector = selectors.activityTxs.createAssetSourceBaseActivity({
      assetName: 'algorand',
      walletAccount: 'exodus_0',
    })
    const result = selector(store.getState())
    expect(result.length).toEqual(1)
    expect(result).toMatchSnapshot()
  })
})
