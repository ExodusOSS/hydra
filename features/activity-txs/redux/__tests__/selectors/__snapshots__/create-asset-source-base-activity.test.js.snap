exports[`createAssetSourceBaseActivity should add indexless orders to activity 1`] = `
{
  "assetName": "flare",
  "date": 2023-09-13T11:59:24.204Z,
  "displayAmount": {
    "type": "NumberUnit",
    "unit": "FLR",
    "unitType": "FLR",
    "value": "7598.14749525",
  },
  "failed": false,
  "id": "swap.flare.eac6bec7798be8",
  "order": {
    "_version": 1,
    "date": "2023-09-13T11:59:24.204Z",
    "displaySvc": "cn",
    "fromAmount": {
      "t": "numberunit",
      "v": {
        "u": {
          "ALGO": 6,
          "microAlgo": 0,
        },
        "v": "1013.41878 ALGO",
      },
    },
    "fromAsset": "algorand",
    "fromTxId": "ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
    "fromWalletAccount": "exodus_0",
    "orderId": "eac6bec7798be8",
    "potentialToTxIds": [],
    "region": "INTL",
    "status": "complete-verified",
    "svc": "cn",
    "synced": true,
    "toAmount": {
      "t": "numberunit",
      "v": {
        "u": {
          "FLR": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
        "v": "7598.14749525 FLR",
      },
    },
    "toAsset": "flare",
    "toTxId": "0x8bbea887472df32167597c36fffc70ac508a01088194e30b877d093fe25f9725",
    "toWalletAccount": "exodus_0",
    "txIds": [
      {
        "txId": "ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
        "txType": "depositTransaction",
      },
      {
        "txId": "0x8bbea887472df32167597c36fffc70ac508a01088194e30b877d093fe25f9725",
        "txType": "payoutTransaction",
      },
    ],
  },
  "pending": false,
  "personalNote": undefined,
  "received": true,
  "sent": false,
  "tx": undefined,
  "txId": "0x8bbea887472df32167597c36fffc70ac508a01088194e30b877d093fe25f9725",
  "type": "exchange",
}
`;

exports[`createAssetSourceBaseActivity should add indexless orders to activity 2`] = `
{
  "assetName": "algorand",
  "date": 2023-09-13T11:59:24.204Z,
  "displayAmount": {
    "type": "NumberUnit",
    "unit": "FLR",
    "unitType": "FLR",
    "value": "7598.14749525",
  },
  "failed": false,
  "id": "swap.algorand.eac6bec7798be8",
  "order": {
    "_version": 1,
    "date": "2023-09-13T11:59:24.204Z",
    "displaySvc": "cn",
    "fromAmount": {
      "t": "numberunit",
      "v": {
        "u": {
          "ALGO": 6,
          "microAlgo": 0,
        },
        "v": "1013.41878 ALGO",
      },
    },
    "fromAsset": "algorand",
    "fromTxId": "ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
    "fromWalletAccount": "exodus_0",
    "orderId": "eac6bec7798be8",
    "potentialToTxIds": [],
    "region": "INTL",
    "status": "complete-verified",
    "svc": "cn",
    "synced": true,
    "toAmount": {
      "t": "numberunit",
      "v": {
        "u": {
          "FLR": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
        "v": "7598.14749525 FLR",
      },
    },
    "toAsset": "flare",
    "toTxId": "0x8bbea887472df32167597c36fffc70ac508a01088194e30b877d093fe25f9725",
    "toWalletAccount": "exodus_0",
    "txIds": [
      {
        "txId": "ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
        "txType": "depositTransaction",
      },
      {
        "txId": "0x8bbea887472df32167597c36fffc70ac508a01088194e30b877d093fe25f9725",
        "txType": "payoutTransaction",
      },
    ],
  },
  "pending": false,
  "personalNote": undefined,
  "received": false,
  "sent": true,
  "tx": {
    "coinAmount": "1013.41878 ALGO",
    "coinName": "algorand",
    "confirmations": 1,
    "currencies": {
      "algorand": {
        "ALGO": 6,
        "microAlgo": 0,
      },
    },
    "date": "2023-09-13T11:59:24.204Z",
    "txId": "ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
    "version": 1,
  },
  "txId": "ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
  "type": "exchange",
}
`;

exports[`createAssetSourceBaseActivity should add orders without txs to activity 1`] = `
{
  "assetName": "flare",
  "date": 2023-09-13T11:59:24.204Z,
  "displayAmount": {
    "type": "NumberUnit",
    "unit": "FLR",
    "unitType": "FLR",
    "value": "10000",
  },
  "failed": false,
  "id": "swap.flare.eac6bec7798be8",
  "order": {
    "_version": 1,
    "date": "2023-09-13T11:59:24.204Z",
    "displaySvc": "cn",
    "fromAmount": {
      "t": "numberunit",
      "v": {
        "u": {
          "ALGO": 6,
          "microAlgo": 0,
        },
        "v": "100 ALGO",
      },
    },
    "fromAsset": "algorand",
    "fromTxId": "algorand-send-tx",
    "fromWalletAccount": "exodus_0",
    "orderId": "eac6bec7798be8",
    "potentialToTxIds": [],
    "region": "INTL",
    "status": "in-progress",
    "svc": "cn",
    "synced": true,
    "toAmount": {
      "t": "numberunit",
      "v": {
        "u": {
          "FLR": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
        "v": "10000 FLR",
      },
    },
    "toAsset": "flare",
    "toTxId": null,
    "toWalletAccount": "exodus_0",
  },
  "pending": true,
  "personalNote": undefined,
  "received": true,
  "sent": false,
  "tx": undefined,
  "txId": null,
  "type": "exchange",
}
`;

exports[`createAssetSourceBaseActivity should add orders without txs to activity 2`] = `
{
  "assetName": "flare",
  "date": 2022-09-13T11:59:24.204Z,
  "displayAmount": {
    "type": "NumberUnit",
    "unit": "FLR",
    "unitType": "FLR",
    "value": "10",
  },
  "failed": false,
  "id": "swap.flare.order-with-tx",
  "order": {
    "_version": 1,
    "date": "2022-09-13T11:59:24.204Z",
    "displaySvc": "cn",
    "fromAmount": {
      "t": "numberunit",
      "v": {
        "u": {
          "ALGO": 6,
          "microAlgo": 0,
        },
        "v": "1 ALGO",
      },
    },
    "fromAsset": "algorand",
    "fromTxId": "tx-with-order-algorand",
    "fromWalletAccount": "exodus_0",
    "orderId": "order-with-tx",
    "potentialToTxIds": [],
    "region": "INTL",
    "status": "complete-verified",
    "svc": "cn",
    "synced": true,
    "toAmount": {
      "t": "numberunit",
      "v": {
        "u": {
          "FLR": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
        "v": "10 FLR",
      },
    },
    "toAsset": "flare",
    "toTxId": "tx-with-order-flare",
    "toWalletAccount": "exodus_0",
  },
  "pending": false,
  "personalNote": undefined,
  "received": true,
  "sent": false,
  "tx": {
    "coinAmount": "10 FLR",
    "coinName": "flare",
    "confirmations": 1,
    "currencies": {
      "flare": {
        "FLR": 18,
        "Gwei": 9,
        "Kwei": 3,
        "Mwei": 6,
        "finney": 15,
        "szabo": 12,
        "wei": 0,
      },
    },
    "date": "2022-09-13T11:59:24.204Z",
    "txId": "tx-with-order-flare",
    "version": 1,
  },
  "txId": "tx-with-order-flare",
  "type": "exchange",
}
`;

exports[`createAssetSourceBaseActivity should group txs with same order 1`] = `
[
  {
    "assetName": "ethereum",
    "date": 2019-07-22T13:54:28.000Z,
    "displayAmount": {
      "type": "NumberUnit",
      "unit": "ETH",
      "unitType": "ETH",
      "value": "3",
    },
    "failed": false,
    "id": "swap.ethereum.btc-eth-order",
    "order": {
      "_version": 1,
      "date": "2019-07-22T13:54:28.000Z",
      "fromAsset": "bitcoin",
      "fromWalletAccount": "exodus_0",
      "orderId": "btc-eth-order",
      "potentialToTxIds": [],
      "status": "",
      "svc": "ss",
      "toAmount": {
        "t": "numberunit",
        "v": {
          "u": {
            "ETH": 18,
            "Gwei": 9,
            "Kwei": 3,
            "Mwei": 6,
            "finney": 15,
            "szabo": 12,
            "wei": 0,
          },
          "v": "3 ETH",
        },
      },
      "toAsset": "ethereum",
      "toWalletAccount": "exodus_0",
      "txIds": [
        {
          "txId": "some-receive-tx",
        },
        {
          "txId": "some-receive-tx-2",
        },
      ],
    },
    "pending": true,
    "personalNote": undefined,
    "received": true,
    "sent": false,
    "tx": {
      "coinAmount": "1 ETH",
      "coinName": "ethereum",
      "confirmations": 1,
      "currencies": {
        "ethereum": {
          "ETH": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
      },
      "date": "2019-07-22T13:54:28.000Z",
      "dropped": false,
      "feeAmount": "0.000189 ETH",
      "feeCoinName": "ethereum",
      "txId": "some-receive-tx",
      "version": 1,
    },
    "txId": "some-receive-tx",
    "type": "exchange",
  },
]
`;

exports[`createAssetSourceBaseActivity should return activity for asset 1`] = `
[
  {
    "assetName": "ethereum",
    "date": 2020-07-22T13:54:28.000Z,
    "failed": false,
    "id": "ethereum.some-receive-tx",
    "pending": false,
    "personalNote": undefined,
    "received": true,
    "sent": false,
    "tx": {
      "coinAmount": "1 ETH",
      "coinName": "ethereum",
      "confirmations": 1,
      "currencies": {
        "ethereum": {
          "ETH": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
      },
      "date": "2020-07-22T13:54:28.000Z",
      "dropped": false,
      "feeAmount": "0.000189 ETH",
      "feeCoinName": "ethereum",
      "txId": "some-receive-tx",
      "version": 1,
    },
    "txId": "some-receive-tx",
    "type": "rx",
  },
  {
    "assetName": "ethereum",
    "date": 2019-07-22T13:54:28.000Z,
    "failed": false,
    "id": "ethereum.some-receive-tx-2",
    "pending": false,
    "personalNote": undefined,
    "received": true,
    "sent": false,
    "tx": {
      "coinAmount": "2 ETH",
      "coinName": "ethereum",
      "confirmations": 1,
      "currencies": {
        "ethereum": {
          "ETH": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
      },
      "date": "2019-07-22T13:54:28.000Z",
      "dropped": false,
      "feeAmount": "0.000189 ETH",
      "feeCoinName": "ethereum",
      "txId": "some-receive-tx-2",
      "version": 1,
    },
    "txId": "some-receive-tx-2",
    "type": "rx",
  },
]
`;

exports[`createAssetSourceBaseActivity should return txs if orders are not supported 1`] = `
[
  {
    "assetName": "algorand",
    "date": 2023-09-13T11:59:24.204Z,
    "failed": false,
    "id": "algorand.ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
    "pending": false,
    "personalNote": undefined,
    "received": true,
    "sent": false,
    "tx": {
      "coinAmount": "1013.41878 ALGO",
      "coinName": "algorand",
      "confirmations": 1,
      "currencies": {
        "algorand": {
          "ALGO": 6,
          "microAlgo": 0,
        },
      },
      "date": "2023-09-13T11:59:24.204Z",
      "txId": "ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
      "version": 1,
    },
    "txId": "ISO7UZYSZAYC2PGTD4OKYJMBFA3JILLKCDSUO3LZB2Q7ECWDYTYQ",
    "type": "rx",
  },
]
`;

exports[`createAssetSourceBaseActivity should use activityIndex for id when available 1`] = `
[
  {
    "assetName": "bitcoin",
    "date": 2019-07-18T12:48:15.000Z,
    "failed": false,
    "id": "bitcoin.bitcoin-tx",
    "pending": false,
    "personalNote": undefined,
    "received": false,
    "sent": true,
    "tx": {
      "coinAmount": "-0.1 BTC",
      "coinName": "bitcoin",
      "confirmations": 1,
      "currencies": {
        "bitcoin": {
          "BTC": 8,
          "bits": 2,
          "satoshis": 0,
        },
      },
      "data": {
        "activityIndex": 0,
        "sent": [
          {
            "address": "btc-address-1",
            "amount": "0.1 BTC",
          },
          {
            "address": "btc-address-2",
            "amount": "0.1 BTC",
          },
        ],
      },
      "date": "2019-07-18T12:48:15.000Z",
      "dropped": false,
      "feeAmount": "0.00002 BTC",
      "feeCoinName": "bitcoin",
      "txId": "bitcoin-tx",
      "version": 1,
    },
    "txId": "bitcoin-tx",
    "type": "tx",
  },
  {
    "assetName": "bitcoin",
    "date": 2019-07-18T12:48:15.000Z,
    "failed": false,
    "id": "bitcoin.bitcoin-tx_1",
    "pending": false,
    "personalNote": undefined,
    "received": false,
    "sent": true,
    "tx": {
      "coinAmount": "-0.1 BTC",
      "coinName": "bitcoin",
      "confirmations": 1,
      "currencies": {
        "bitcoin": {
          "BTC": 8,
          "bits": 2,
          "satoshis": 0,
        },
      },
      "data": {
        "activityIndex": 1,
        "sent": [
          {
            "address": "btc-address-1",
            "amount": "0.1 BTC",
          },
          {
            "address": "btc-address-2",
            "amount": "0.1 BTC",
          },
        ],
      },
      "date": "2019-07-18T12:48:15.000Z",
      "dropped": false,
      "feeAmount": "0.00002 BTC",
      "feeCoinName": "bitcoin",
      "txId": "bitcoin-tx",
      "version": 1,
    },
    "txId": "bitcoin-tx",
    "type": "tx",
  },
]
`;
